### PR TITLE
Allow jobs to be deleted from FMS via rpc

### PR DESCRIPTION
**Pull request by eutopian**
Currently there is no straightforward way for FMS to send job deletion requests to managed node ops other than having the node ops manually search for the job via external ID. This will allow jobs proposed and managed by FMS to also be deleted from FMS.

### CODE REVIEW

File: core/services/feeds/models.go [hunk 1]
```diff
@@ -167,6 +166,9 @@ const (
 	// but cancelled by the node op. A cancelled spec is not being run by the
 	// node.
 	SpecStatusCancelled SpecStatus = "cancelled"
+	// SpecStatusDeleted defines a spec status which was deleted by the FMS.
+	// A deleted job spec is not being run by the node.
+	SpecStatusDeleted SpecStatus = "deleted"
```
Comment (jkongie) [inline core/services/feeds/models.go:171]: I don't believe you need to add a spec status for deleted. It is only the job proposal that is deleted

File: core/store/migrate/migrations/0158_update_job_proposal_spec_job_proposal_statuses.sql [hunk 2]
```diff
@@ -0,0 +1,144 @@
+-- +goose Up
+-- +goose StatementBegin
+-- JobProposals Table
+-- We must remove the old contraint to add an enum value to support Postgres v11
+ALTER TABLE job_proposals DROP CONSTRAINT chk_job_proposals_status_fsm;
```
Comment (jkongie) [inline core/store/migrate/migrations/0158_update_job_proposal_spec_job_proposal_statuses.sql:5]: Since you are only adding you can do this now since PG 9.1

`ALTER TYPE enum_type ADD VALUE 'new_value';`

File: core/store/migrate/migrations/0158_update_job_proposal_spec_job_proposal_statuses.sql [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+		)
+	);
+
+-- JobProposalSpecs Table
+-- We must remove the old constraint to add an enum value to support Postgres v11
+DROP INDEX idx_job_proposal_specs_job_proposal_id_and_status;
+
+-- Drop the cancelled enum value. Unfortunately postgres does not support a
+-- a way to remove a value from an enum.
+ALTER TYPE job_proposal_spec_status
```
Comment (jkongie) [inline core/store/migrate/migrations/0158_update_job_proposal_spec_job_proposal_statuses.sql:60]: You shouldn't need to add any new spec status

Comment (eutopian) [reply core/store/migrate/migrations/0158_update_job_proposal_spec_job_proposal_statuses.sql:60]: was kind of going back and forth with this... yeah, I think cancelling them is fine, altho the `CancelSpec` orm method also cancels the proposal. I guess it's probably fine leaving them as approved if the proposal itself is deleted.

File: core/store/migrate/migrations/0158_update_job_proposal_spec_job_proposal_statuses.sql (hunk 2 above)
Comment (eutopian) [reply core/store/migrate/migrations/0158_update_job_proposal_spec_job_proposal_statuses.sql:5]: So used to the rename recreate reassign drop flow of updating a PG enum

Comment (eutopian) [reply core/store/migrate/migrations/0158_update_job_proposal_spec_job_proposal_statuses.sql:5]: Unfortunately, this won't allow me to update the null constraint check in the same migration as I get `Unsafe use of new value of enum type` error for trying to use the new constraint in the same transaction.

File: core/services/feeds/models.go [hunk 4]
```diff
@@ -150,11 +151,10 @@ type JobProposal struct {
 	UpdatedAt      time.Time
 }
 
-// SpecStatus is the status of each proposed spec.
```
Comment (jkongie) [inline core/services/feeds/models.go:153]: Why remove this comment?

File: core/store/migrate/migrations/0158_update_job_proposal_status.sql [hunk 5]
```diff
@@ -0,0 +1,93 @@
+-- +goose Up
+-- +goose StatementBegin
+-- JobProposals Table
+-- We must remove the old contraint to add an enum value to support Postgres v11
+ALTER TABLE job_proposals DROP CONSTRAINT chk_job_proposals_status_fsm;
```
Comment (jkongie) [inline core/store/migrate/migrations/0158_update_job_proposal_status.sql:5]: You can add new enum types since Postgres 9.1 without needing to create a new type.

File: core/services/feeds/service.go [hunk 6]
```diff
@@ -524,6 +641,10 @@ func (s *service) RejectSpec(ctx context.Context, id int64) error {
 		return errors.Wrap(err, "orm: job proposal")
 	}
 
+	if proposal.Status == JobProposalStatusDeleted {
```
Comment (jkongie) [inline core/services/feeds/service.go:644]: I don't think you need this. I don't see ay harm in allowing them to reject a spec for deleted job proposal

File: core/store/migrate/migrations/0158_update_job_proposal_status.sql (hunk 5 above)
Comment (eutopian) [reply core/store/migrate/migrations/0158_update_job_proposal_status.sql:5]: response got lost in the long history; Unfortunately, this won't allow me to update the null constraint check in the same migration as I get Unsafe use of new value of enum type error for trying to use the new constraint in the same transaction.

File: core/services/feeds/models.go (hunk 4 above)
Comment (eutopian) [reply core/services/feeds/models.go:153]: accident

File: core/services/feeds/orm.go [hunk 7]
```diff
@@ -484,19 +485,29 @@ RETURNING job_proposal_id;
...
-	if err := o.q.WithOpts(qopts...).Get(&jpID, stmt, JobProposalStatusCancelled, id); err != nil {
+	if err := o.q.WithOpts(qopts...).Get(&jpID, stmt, SpecStatusCancelled, id); err != nil {
 		return err
 	}
 
 	stmt = `
 UPDATE job_proposals
-SET status = $1,
-	external_job_id = $2,
+SET STATUS = subquery.updateStatus,
```
Comment (jkongie) [inline core/services/feeds/orm.go:494]: Why did you need uppercase `status`?

File: core/services/feeds/orm.go [hunk 8]
```diff
@@ -614,21 +654,27 @@ RETURNING job_proposal_id;
...
 	var jpID int64
-	if err := o.q.WithOpts(qopts...).Get(&jpID, stmt, JobProposalStatusRejected, id); err != nil {
+	if err := o.q.WithOpts(qopts...).Get(&jpID, stmt, SpecStatusRejected, id); err != nil {
 		return err
 	}
 
 	stmt = `
 UPDATE job_proposals
-SET status = subquery.updateStatus,
+SET STATUS = subquery.updateStatus,
```
Comment (jkongie) [inline core/services/feeds/orm.go:663]: Why did you need uppercase `status`?

File: core/services/feeds/orm_test.go [hunk 9]
```diff
@@ -731,6 +731,80 @@ func Test_ORM_ApproveSpec(t *testing.T) {
...
+				actualJP, err := orm.GetJobProposal(jpID)
+				require.NoError(t, err)
+
+				assert.Equal(t, tc.wantProposalStatus, actualJP.Status)
+				assert.False(t, actualJP.PendingUpdate)
+			}
+		})
+	}
+}
+
```
Comment (jkongie) [inline core/services/feeds/orm_test.go:804]: Need to update the `RejectSpec` test

File: core/services/feeds/orm.go [hunk 10]
```diff
@@ -541,6 +552,35 @@ SELECT exists (
...
 
+// DeleteProposal performs a soft delete of the job proposal by setting the
+// status to deleted, remove the external job id, and update the status to
+// deleted
+func (o *orm) DeleteProposal(id int64, qopts ...pg.QOpt) error {
+	stmt := `
+UPDATE job_proposals
+SET status = $1,
+	external_job_id = $2,
+	pending_update = TRUE,
```
Comment (jkongie) [inline core/services/feeds/orm.go:563]: `pending_update` should be set to false

File: core/services/feeds/service_test.go [hunk 11]
```diff
@@ -71,7 +71,7 @@ const OCR2TestSpec = `
 type               = "offchainreporting2"
 pluginType         = "median"
 schemaVersion      = 1
-relay              = "evm"
+relay              = "evm"e
```
Comment (jkongie) [inline core/services/feeds/service_test.go:74]: typo?

File: core/store/migrate/migrations/0161_update_job_proposal_status.sql [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+	'pending',
+	'approved',
+	'rejected',
+	'cancelled',
+	'deleted',
+	'revoked'
+);
+
+ALTER TABLE job_proposals
+ALTER COLUMN STATUS TYPE job_proposal_status USING STATUS::text::job_proposal_status;
```
Comment (jkongie) [inline core/store/migrate/migrations/0161_update_job_proposal_status.sql:20]: Better readability 

```suggestion
ALTER COLUMN status TYPE job_proposal_status USING STATUS::text::job_proposal_status;
```

File: core/store/migrate/migrations/0161_update_job_proposal_status.sql [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+-- a way to remove a value from an enum.
+ALTER TYPE job_proposal_status
+RENAME TO job_proposal_status_old;
+
+CREATE TYPE job_proposal_status AS ENUM('pending', 'approved', 'rejected', 'cancelled');
+
+-- This will fail if any records are using the 'deleted' or 'revoked' enum.
+-- Manually update these as we cannot decide what you want to do with them.
+ALTER TABLE job_proposals
+ALTER COLUMN STATUS TYPE job_proposal_status USING STATUS::text::job_proposal_status;
```
Comment (jkongie) [inline core/store/migrate/migrations/0161_update_job_proposal_status.sql:68]:
```suggestion
ALTER COLUMN status TYPE job_proposal_status USING status::text::job_proposal_status;
```

**Review by jkongie — COMMENTED** [review]
Some minor changes but it looks good

File: core/services/feeds/orm.go (hunk 8 above)
Comment (eutopian) [reply core/services/feeds/orm.go:663]: it was a sql formatter editor on save

File: core/services/feeds/orm.go (hunk 10 above)
Comment (eutopian) [reply core/services/feeds/orm.go:563]: I set the pending_update to true here because they need to go cancel the job

File: core/store/migrate/migrations/0161_update_job_proposal_status.sql (hunk 12 above)
Comment (eutopian) [reply core/store/migrate/migrations/0161_update_job_proposal_status.sql:20]: My sql formatter changed this on save, but I guess status is fine in postgres but a keyword in mysql

File: core/services/feeds/service.go [hunk 14]
```diff
@@ -406,6 +407,45 @@ func (s *service) ListJobProposalsByManagersIDs(ids []int64) ([]JobProposal, err
...
+			return 0, errors.New("cannot delete a job proposal belonging to another feeds manager")
+		}
+	}
+
+	pctx := pg.WithParentCtx(ctx)
+	if txerr := s.orm.DeleteProposal(proposal.ID, pctx); txerr != nil {
+		s.lggr.Errorw("Failed to delete the proposal", "error", txerr)
+
+		return 0, errors.Wrap(txerr, "DeleteProposal failed")
+	}
```
Comment (jkongie) [inline core/services/feeds/service.go:440]: You can just use `err` here

 ```suggestion
	if err = s.orm.DeleteProposal(proposal.ID, pctx); err != nil {
		s.lggr.Errorw("Failed to delete the proposal", "error", err)

		return 0, errors.Wrap(err, "DeleteProposal failed")
	}
```

File: core/services/feeds/service.go [hunk 15]
```diff
@@ -406,6 +407,45 @@ func (s *service) ListJobProposalsByManagersIDs(ids []int64) ([]JobProposal, err
...
+
+	if err != nil {
```
Comment (jkongie) [inline core/services/feeds/service.go:442]: This err check seem redundant

File: core/services/feeds/service_test.go [hunk 16]
```diff
@@ -632,6 +632,92 @@ func Test_Service_ProposeJob(t *testing.T) {
...
+					Return(&feeds.JobProposal{
+						FeedsManagerID: 2,
+						RemoteUUID:     approved.RemoteUUID,
+						Status:         feeds.JobProposalStatusApproved,
+					}, nil)
+			},
+			args:    args,
+			wantErr: "cannot delete a job proposal belonging to another feeds manager",
+		},
+		{
```
Comment (jkongie) [inline core/services/feeds/service_test.go:685]: Missing a test on `GetJobProposalByRemoteUUID` no rows  error

**Review by jkongie — CHANGES_REQUESTED** [review]
Some minor changes but otherwise looks good

**Review by jkongie — APPROVED** [review]
